File: DSA/Ayush_Mishra_590014844/Day_36_Question_2.c
#include <stdio.h>
#include <string.h>

#define MAX 100

char stack[MAX];
int top = -1;

int is_match(char open, char close) {
    return (open == '(' && close == ')') ||
           (open == '[' && close == ']') ||
           (open == '{' && close == '}');
}

int main() {
    char expr[MAX];
    printf("Enter an expression: ");
    scanf("%s", expr);

    int len = strlen(expr);
    for (int i = 0; i < len; i++) {
        if (expr[i] == '(' || expr[i] == '[' || expr[i] == '{')
            stack[++top] = expr[i];
        else if (expr[i] == ')' || expr[i] == ']' || expr[i] == '}') {
            if (top == -1 || !is_match(stack[top--], expr[i])) {
                printf("Not Balanced\n");
                return 0;
            }
        }
    }
    if (top == -1)
        printf("Balanced\n");
    else
        printf("Not Balanced\n");

    return 0;
}

File: DSA/Ayush_Mishra_590014844/Day_36_Question_2.py
def is_balanced(expr):
    stack = []
    match = {')': '(', ']': '[', '}': '{'}
    for ch in expr:
        if ch in '([{':
            stack.append(ch)
        elif ch in ')]}':
            if not stack or stack[-1] != match[ch]:
                return False
            stack.pop()
    return len(stack) == 0

expr = input("Enter an expression: ")
print("Balanced" if is_balanced(expr) else "Not Balanced")
